File: python/publish_sub.py
# coding=utf8
'''
发布／订阅
Publish/Subscribe
subscribe
https://github.com/rabbitmq/rabbitmq-tutorials
https://www.rabbitmq.com/tutorials/tutorial-three-python.html
          /[][][]
(P) -> {X}
          \[][][]
'''

import pika

# 认证信息
credentials = pika.PlainCredentials('admin', 'admin')

connection = pika.BlockingConnection(
    pika.ConnectionParameters(host='192.168.50.83', credentials=credentials))

channel = connection.channel()
# 创建一个fanout类型的交换机，命名为logs
# 扇型交换机（fanout exchange），它把消息发送给它所知道的所有队列
channel.exchange_declare(exchange='logs', exchange_type='fanout')

result = channel.queue_declare(queue='', exclusive=True)
queue_name = result.method.queue

channel.queue_bind(exchange='logs', queue=queue_name)


def callback(ch, method, properties, body):
    print('subscribe:{}'.format(body))

channel.basic_consume(
    queue=queue_name,
    on_message_callback=callback,
    auto_ack=True
)

channel.start_consuming()

File: php/publish_sub.php
<?php

/**
 * rabbitmq
 * 发布／订阅
 * Publish/Subscribe
 * received
 * https://github.com/rabbitmq/rabbitmq-tutorials
 * https://www.rabbitmq.com/tutorials/tutorial-three-php.html
 * 
 * https://xiaoxiami.gitbook.io/rabbitmq_into_chinese_php/ying-yong-jiao-cheng/php-ban/3-publish_subscribe.md
 
          /[][][]
(P) -> {X}
          \[][][]

 */

defined('DS') or define('DS', DIRECTORY_SEPARATOR);
require_once __DIR__. DS . 'vendor' .DS.'autoload.php';

use PhpAmqpLib\Connection\AMQPStreamConnection;

$connection = new AMQPStreamConnection('192.168.50.83', 5672, 'admin', 'admin');
$channel = $connection->channel();

# 创建一个fanout类型的交换机，命名为logs
# 扇型交换机（fanout），它把消息发送给它所知道的所有队列
$channel->exchange_declare('logs', 'fanout', false, false, false);

# 定义临时队列
# 当与消费者（consumer）断开连接的时候，这个队列被立即删除
list($queue_name, ,) = $channel->queue_declare('', false, false, true, false);

# logs交换机将会把消息添加到我们的队列中
$channel->queue_bind($queue_name, 'logs');


$callback = function($msg){
    echo 'Subscribe:', $msg->body, PHP_EOL;
};

$channel->basic_consume($queue_name, '', false, true, false, false, $callback);

while($channel->is_consuming()){
    $channel->wait();
}

$channel->close();
$connection->close();
